Add testing instructions to the README (#647)
microsoft/STL: README.md

because they need to be updated whenever source files are added/renamed/deleted. We'll delete the legacy machinery as
soon as possible.)

* Tests: **In progress.** We rely on three test suites: std, tr1, and [libcxx][]. We've partially ported std and tr1,

> and fully ported libcxx to run under [lit][] using the various configuration/compilers we test internally.

Reviewer: ```suggestion
and fully ported libcxx to run under [lit][] using the various configurations/compilers we test internally.
```

because they need to be updated whenever source files are added/renamed/deleted. We'll delete the legacy machinery as
soon as possible.)

> * Tests: **In progress.** We rely on three test suites: std, tr1, and [libcxx][]. We've partially ported std and tr1,

Author: The `/clr:pure` and `/BE` tests aren't running.

UNSUPPORTED: tr1 :: tests/cwchar1:23 (13 of 28)
UNSUPPORTED: tr1 :: tests/cwchar1:24 (14 of 28)
PASS: tr1 :: tests/cwchar1:07 (15 of 28)
PASS: tr1 :: tests/cwchar1:08 (16 of 28)
PASS: tr1 :: tests/cwchar1:10 (17 of 28)
PASS: tr1 :: tests/cwchar1:16 (18 of 28)
PASS: tr1 :: tests/cwchar1:17 (19 of 28)
PASS: tr1 :: tests/cwchar1:14 (20 of 28)
PASS: tr1 :: tests/cwchar1:12 (21 of 28)
PASS: tr1 :: tests/cwchar1:13 (22 of 28)
PASS: tr1 :: tests/cwchar1:19 (23 of 28)
PASS: tr1 :: tests/cwchar1:18 (24 of 28)
PASS: tr1 :: tests/cwchar1:15 (25 of 28)
PASS: tr1 :: tests/cwchar1:25 (26 of 28)
PASS: tr1 :: tests/cwchar1:26 (27 of 28)
PASS: tr1 :: tests/cwchar1:27 (28 of 28)

Testing Time: 3.96s
  Expected Passes    : 23
  Unsupported Tests  : 5
```

In the above example we see that twenty-three tests succeeded and five were unsupported.

### Result Code Values

Our tests use the standard [lit result codes][] and the addition of a non-standard result code `SKIP`. For our tests
one need mostly only concern themselves with the `PASS`, `XFAIL`, `XPASS`, `FAIL`, and `UNSUPPORTED` result codes.

The `PASS` and `FAIL` result codes are self explanatory. We want our tests to `PASS` and not `FAIL`.

> The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successfull result and indicates that we expected the

Reviewer 2: ```suggestion
The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successful result and indicates that we expected the

```

    msvcp140d_oss.dll
```

# How To Run The Tests From The Developer Command Prompt For VS

1. Follow steps 1-9 of

> [How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt).

Reviewer: Can these be the cool `[meow][]` form, or does that not work with internal section links?

Author: Maybe it does? I'll try it.

Author: Looks like it does.

2. Invoke `git submodule update --init llvm-project`
3. Invoke `cmake -G Ninja -DCMAKE_TOOLCHAIN_FILE={where your vcpkg clone is located}\scripts\buildsystems\vcpkg.cmake
-DENABLE_TESTS=TRUE -S . -B {wherever you want binaries}`. This differs from above only in `-DENABLE_TESTS=TRUE`.
4. If you have already followed the steps from
[How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt), and have not

> changed the value of {wherever you want binaries} in step four, then there is no need to rebuild to run the tests.

Reviewer: ```suggestion
changed the value of `{wherever you want binaries}` in step 4, then there is no need to rebuild to run the tests.
```

In addition to following the above steps you must also have [Python][] 3.8 or newer, and have LLVM's `bin` directory on
the PATH.

## Running All The Tests

After configuring and building the project, running `ctest` from the build output directory will run all the tests.
CTest will only display the standard error output of tests that failed. In order to get more details from CTest's
`lit` invocations, run the tests with `ctest -V`.

## Running A Subset Of The Tests

`${PROJECT_BINARY_DIR}\tests\llvm-lit\llvm-lit.py` can be invoked on a subdirectory of a testsuite and will execute all
the tests under that subdirectory. This can mean executing the entirety of a single testsuite, running all tests under
a category in libcxx, or running a single test in `std` and `tr1`.

## Examples

```
:: This command configures the project with tests enabled. It assumes you are using the vcpkg submodule and have
:: already installed boost. It also assumes you have inited and updated the llvm-project submodule.
C:\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake^

>  -DENABLE_TESTS=TRUE ..

Reviewer: I think it would be much less confusing to put the space before the caret:
```suggestion
C:\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ^
-DENABLE_TESTS=TRUE ..
```

> :: As stated above this step is only strictly necessary if you have yet to build the STL or if you have changed the

Reviewer: ```suggestion
:: As stated above, this step is only strictly necessary if you have yet to build the STL or if you have changed the
```

:: output directory of the binaries. Any changes or additions in any of the existing testsuites do not require
:: recompilation or reconfiguration to take effect when running the tests.
C:\STL\build>ninja

:: This command will run all of the testsuites with verbose output.
C:\STL\build>ctest -V

> :: This command will run all of the tr1 testuite

Reviewer: ```suggestion
:: This command will run all of the tr1 testsuite.
```

> C:\STL\build>ctest -R tr1

Reviewer: Should the two examples with `tr1` instead mention `std` which is the primary test suite that people should be interested in?

Author: Sure I was just trying to have at least one example of each. Makes sense though.

> :: This command will also run all of the tr1 testuite.

Reviewer: ```suggestion
:: This command will also run all of the tr1 testsuite.
```

C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\tr1

:: If you want to run a subset of a testsuite you need to point it to the right place in the sources. The following
:: will run the single test found under VSO_0000000_any_calling_conventions.
C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions

> :: You can invoke llvm-lit with any arbitrary subdirectory of a testuite. In libcxx this allows you to have a finer

Reviewer: ```suggestion
:: You can invoke llvm-lit with any arbitrary subdirectory of a testsuite. In libcxx this allows you to have finer
```

:: control over what category of tests you would like to run. The following will run all the libcxx map tests.
C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\llvm-project\libcxx\test\std\containers\associative\map
```

> ## Interpreting The Results Of Tests

Reviewer: It seems confusing that this is the same header level as its "children" `CTest` and `llvm-lit` below.

Author: Woops, the next two were supposed to be nested.

## CTest

When running the tests via CTest each of the testsuites is considered to be a single test. If any single test in a
testsuite fails, CTest will report the test which represents that testsuite as failed.

Example:
```
67% tests passed, 1 tests failed out of 3

Total Test time (real) = 2441.55 sec

The following tests FAILED:
      1 - libcxx (Failed)
```

CTest will output everything that was sent to stderr for each of the failed testsuites, which can be used to identify
which individual test within the testsuite failed. It can sometimes be helpful to run CTest with the `-V` option in
order to see the stdout of the tests.

## llvm-lit

When running the tests directly via the generated `llvm-lit.py` script the result of each test will be printed. The

> format of each result is {[Result Code](#result-code-values)}: {Testsuite Name} :: {Test Name}:{Configuration Number}.

Reviewer: ```suggestion
format of each result is `{[Result Code](#result-code-values)}: {Testsuite Name} :: {Test Name}:{Configuration Number}`.
```

Author: I tried that, you can't have links inside of a code snipet.

Author: Maybe I should drop the link?

PASS: tr1 :: tests/cwchar1:02 (3 of 28)
PASS: tr1 :: tests/cwchar1:03 (4 of 28)
PASS: tr1 :: tests/cwchar1:00 (5 of 28)
PASS: tr1 :: tests/cwchar1:04 (6 of 28)
PASS: tr1 :: tests/cwchar1:05 (7 of 28)
PASS: tr1 :: tests/cwchar1:09 (8 of 28)
PASS: tr1 :: tests/cwchar1:06 (9 of 28)
UNSUPPORTED: tr1 :: tests/cwchar1:20 (10 of 28)
UNSUPPORTED: tr1 :: tests/cwchar1:21 (11 of 28)
UNSUPPORTED: tr1 :: tests/cwchar1:22 (12 of 28)
UNSUPPORTED: tr1 :: tests/cwchar1:23 (13 of 28)
UNSUPPORTED: tr1 :: tests/cwchar1:24 (14 of 28)
PASS: tr1 :: tests/cwchar1:07 (15 of 28)
PASS: tr1 :: tests/cwchar1:08 (16 of 28)
PASS: tr1 :: tests/cwchar1:10 (17 of 28)
PASS: tr1 :: tests/cwchar1:16 (18 of 28)
PASS: tr1 :: tests/cwchar1:17 (19 of 28)
PASS: tr1 :: tests/cwchar1:14 (20 of 28)
PASS: tr1 :: tests/cwchar1:12 (21 of 28)
PASS: tr1 :: tests/cwchar1:13 (22 of 28)
PASS: tr1 :: tests/cwchar1:19 (23 of 28)
PASS: tr1 :: tests/cwchar1:18 (24 of 28)
PASS: tr1 :: tests/cwchar1:15 (25 of 28)
PASS: tr1 :: tests/cwchar1:25 (26 of 28)
PASS: tr1 :: tests/cwchar1:26 (27 of 28)
PASS: tr1 :: tests/cwchar1:27 (28 of 28)

Testing Time: 3.96s
  Expected Passes    : 23
  Unsupported Tests  : 5
```

> In the above example we see that twenty-three tests succeeded and five were unsupported.

Reviewer: ```suggestion
In the above example we see that 23 tests succeeded and 5 were unsupported.
```

### Result Code Values

> Our tests use the standard [lit result codes][] and the addition of a non-standard result code `SKIP`. For our tests

Reviewer: "Our tests use ... the addition" sounds strange. "and a non-standard result code" or "and an additional non-standard result code" would sound better.

> Our tests use the standard [lit result codes][] and the addition of a non-standard result code `SKIP`. For our tests

Reviewer: The explanation of `SKIP` is skipped.

> one need mostly only concern themselves with the `PASS`, `XFAIL`, `XPASS`, `FAIL`, and `UNSUPPORTED` result codes.

Reviewer: "one need mostly only concern themselves" is awkward (this document doesn't attempt to avoid "you" elsewhere).

Can this be phrased more simply as "For our tests, only the BLAH result codes are relevant."?

> The `PASS` and `FAIL` result codes are self explanatory. We want our tests to `PASS` and not `FAIL`.

Reviewer: ```suggestion
The `PASS` and `FAIL` result codes are self-explanatory. We want our tests to `PASS` and not `FAIL`.
```

> The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successful result and indicates that we expected the

Reviewer: ```suggestion
The `XPASS` and `XFAIL` result codes are less obvious. `XFAIL` is actually a successful result and indicates that we expected the
```
Note that this will overflow 120 columns.

> The `XPASS` and `XFAIL` are less obvious. `XFAIL` is actually a successful result and indicates that we expected the

Reviewer: I find it mildly jarring that XPASS and XFAIL are initially mentioned in that order, then explained in reverse order. I would expect XPASS to be explained before XFAIL. (Note that mentioning XFAIL before XPASS would create some inconsistency with the initial explanation of PASS and FAIL.)

> 

Reviewer: Spurious empty line.

test to fail and it did. `XPASS` is a failure result and indicates that we expected a test to fail but it passed.
Typically this means that the `expected_results.txt` file for the testsuite needs to be modified and a `FAIL` entry
needs to be removed.

The `UNSUPPORTED` result code means that the requirements for a test are not met and so it will not be run. Currently

> all tests which use the `/BE` or `clr:pure` flags are unsupported.

Reviewer: ```suggestion
all tests which use the `/BE` or `/clr:pure` options are unsupported.
```

In addition to following the above steps you must also have [Python][] 3.8 or newer, and have LLVM's `bin` directory on
the PATH.

## Running All The Tests

After configuring and building the project, running `ctest` from the build output directory will run all the tests.
CTest will only display the standard error output of tests that failed. In order to get more details from CTest's
`lit` invocations, run the tests with `ctest -V`.

## Running A Subset Of The Tests

`${PROJECT_BINARY_DIR}\tests\llvm-lit\llvm-lit.py` can be invoked on a subdirectory of a testsuite and will execute all
the tests under that subdirectory. This can mean executing the entirety of a single testsuite, running all tests under
a category in libcxx, or running a single test in `std` and `tr1`.

## Examples

```
:: This command configures the project with tests enabled. It assumes you are using the vcpkg submodule and have
:: already installed boost. It also assumes you have inited and updated the llvm-project submodule.
C:\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ^

> -DENABLE_TESTS=TRUE ..

Reviewer 3: I think these should have a newline between the comments and the commands. It's a bit clustered and hard to parse otherwise.

Author: Sure, I can see what it looks like.

PASS: tr1 :: tests/cwchar1:16 (18 of 28)
PASS: tr1 :: tests/cwchar1:17 (19 of 28)
PASS: tr1 :: tests/cwchar1:14 (20 of 28)
PASS: tr1 :: tests/cwchar1:12 (21 of 28)
PASS: tr1 :: tests/cwchar1:13 (22 of 28)
PASS: tr1 :: tests/cwchar1:19 (23 of 28)
PASS: tr1 :: tests/cwchar1:18 (24 of 28)
PASS: tr1 :: tests/cwchar1:15 (25 of 28)
PASS: tr1 :: tests/cwchar1:25 (26 of 28)
PASS: tr1 :: tests/cwchar1:26 (27 of 28)
PASS: tr1 :: tests/cwchar1:27 (28 of 28)

Testing Time: 3.96s
  Expected Passes    : 23
  Unsupported Tests  : 5
```

In the above example we see that 23 tests succeeded and 5 were unsupported.

### Result Code Values

Our tests use the standard [lit result codes][], and a non-standard result code: `SKIP`. For our tests, only the
`PASS`, `XFAIL`, `XPASS`, `FAIL`, and `UNSUPPORTED` standard result codes are relevant.

The `PASS` and `FAIL` result codes are self-explanatory. We want our tests to `PASS` and not `FAIL`.

The `XPASS` and `XFAIL` result codes are less obvious. `XPASS` is actually a failure result and indicates that we
expected a test to fail but it passed. `XFAIL` is a successful result and indicates that we expected the test to fail

> and it did. Typically an `XPASS` result means that the `expected_results.txt` file for the testsuite needs to be

Reviewer 4: I gues you want a negation here `and it did *not`

Author: No I think what is written is correct. `XFAIL` means that we expected the test to fail and it did. `XPASS` means we expected the test to fail but it passed.

I know, it's a little confusing.

Reviewer 4: The full sentence is:

>`XFAIL` is a successful result and indicates that we expected the test to fail
and it did.

So we expected the test to fail and it did?

:: output directory of the binaries. Any changes or additions in any of the existing testsuites do not require
:: recompilation or reconfiguration to take effect when running the tests.
C:\STL\build>ninja

:: This command will run all of the testsuites with verbose output.
C:\STL\build>ctest -V

:: This command will run all of the tr1 testuite
C:\STL\build>ctest -R tr1

:: This command will also run all of the tr1 testuite.
C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\tr1

:: If you want to run a subset of a testsuite you need to point it to the right place in the sources. The following
:: will run the single test found under VSO_0000000_any_calling_conventions.
C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions

:: You can invoke llvm-lit with any arbitrary subdirectory of a testuite. In libcxx this allows you to have a finer
:: control over what category of tests you would like to run. The following will run all the libcxx map tests.
C:\STL\build>python tests\llvm-lit\llvm-lit.py ..\llvm-project\libcxx\test\std\containers\associative\map
```

> 

Reviewer 2: Should we have another section describing how to interpret the results?

Author: Sure I guess that's worth it since `ctest` has different output than `llvm-lit`.

Reviewer 2: ...and our user may be unfamiliar with both.

    msvcp140d_oss.dll
```

# How To Run The Tests

In order to run any of the tests you must be in a correctly configured environment for the architecture you are

> targeting, have python 3.8 or newer and have `clang-cl` on the PATH. If you are testing the x64 flavor of the STL then

Reviewer 2: We shouldn't assume that every user knows what `clang-cl` is or where it comes from.
```suggestion
targeting, have python 3.8 or newer and have LLVM's `bin` directory on the PATH. If you are testing the x64 flavor of the STL then
```

running from the "x64 Native Tools Command Prompt" is sufficient for setting up the environment.

## Running All The Tests

> After configuring the project, running `ctest` from the root of the build output directory will run all the tests.

Reviewer 2: Do we really mean "After building the project"? 

Directories don't have roots.
```suggestion
After configuring the project, running `ctest` from the build output directory will run all the tests.
```

> CTest will only display the standard error output of tests which failed. In order to get more details from CTest's

Reviewer: "which" is nonrestrictive, "that" is restrictive, so this should be:

```suggestion
CTest will only display the standard error output of tests that failed. In order to get more details from CTest's
```

Reviewer 2: This doesn't look resolved to me.

Author: Thanks for catching that. I usually click resolved as I fix suggestions. Misaligned on this one I guess.

`lit` invocations, run the tests with `ctest -V`.

## Running A Subset Of The Tests

In order to run a subset of one of the testsuites you must invoke `lit` yourself and point it to the desired
subdirectory of tests.

The CMake configuration step produces a Python script, `${PROJECT_BINARY_DIR}\tests\llvm-lit\llvm-lit.py`.
`llvm-lit.py` simply calls into the standard `lit` found in `llvm-project\llvm\utils\lit` with certain options
pre-configured. Namely a mapping of `lit.cfg` files to `lit.site.cfg` files is provided, which informs lit which
site-specific config file to load whenever it discovers a general config file.

> While one can directly use `lit` to run our tests by hand, we only support using the generated `llvm-lit`.

Reviewer: I am confused by this entire section. The first paragraph said "you must invoke `lit` yourself", but the third paragraph says that directly using `lit` is unsupported, and all of the examples demonstrate `llvm-lit.py`.

Additionally, users of the test infrastructure shouldn't care about how information is communicated to the underlying `lit`. (This seems like it would be more appropriate as comments in the test infrastructure, or a README in a subdirectory of the test infrastructure.)

Should this section simply explain that `llvm-lit.py` can be invoked with a test suite directory (like `tests\std`), a test category directory (e.g. libcxx's nicely organized subdirectories), or an individual test directory?

Author: I think I'll add a little bit more context in the examples and see what you think.

> ### Examples

Reviewer: I think should be given equal weight with the preceding subsections:
```suggestion
## Examples
```

```

> C:\Dev\STL\build>cmake -GNinja -DCMAKE_CXX_COMPILER=cl -DCMAKE_TOOLCHAIN_FILE=..\vcpkg\scripts\buildsystems\vcpkg.cmake ..

Reviewer: Should a comment explain what this line is doing?

> 

Reviewer: Should there be an example of running all of the tests with `ctest`?

:: This command will run all of the std testsuite.

> C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py tests\std

Reviewer: These commands don't appear to have been tested. This line says `tests\std` but the following line says `..\tests\std`; I suspect that the latter is correct.

Author: No both lines work. If you point to `tests\std` there is a `lit.site.cfg` file that gets placed there and will load all the config as necessary for running the entire `std` testsuite. However, if you want to run tests under a specific directory you need to point to the actual source directory.

Author: I'll be consistent and not use examples where the user points to something in the build directory even if they can.

> :: However if you want to run a subset of std you need to point it to the right place in the sources.

Reviewer: ```suggestion
:: However, if you want to run a subset of std, you need to point it to the right place in the sources.
```
Or perhaps this should talk about running all configurations of one test in std?

C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py ..\tests\std\tests\VSO_0000000_any_calling_conventions

:: The same applies for libcxx. The following will run all the libcxx containers tests.

> C:\Dev\STL\build>python tests\llvm-lit\llvm-lit.py ..\llvm-project\libcxx\test\std\containers

Reviewer: Did we ever tell users to init/update the `llvm-project` submodule? (Or, separately, the `vcpkg` submodule?)

Author: Something I hadn't considered in my test harness changes is that previously there was no hard requirement to init the llvm-project submodule in order to configure and build the project. Now there is. I should probably go fix that.

# How To Run The Tests From The Developer Command Prompt For VS

1. Follow steps 1-9 of
[How To Build With A Native Tools Command Prompt](#how-to-build-with-a-native-tools-command-prompt).
2. Invoke `git submodule init llvm-project`

> 3. Invoke `git submodule update`

Reviewer 2: ```suggestion
2. Invoke `git submodule update --init llvm-project`
```

and it did. Typically an `XPASS` result means that the `expected_results.txt` file for the testsuite needs to be
modified. If the `XPASS` result is a test legitimately passing, the usual course of action would be to remove a `FAIL`
entry from the `expected_results.txt`. However, some tests from `libcxx` mark themselves as `XFAIL` (meaning they
expect to fail) for features they have added tests for but have yet to implement in `libcxx`. If the STL implements
those features first the tests will begin passing unexpectedly for us and return `XPASS` results. In order to resolve
this it is necessary to add a `PASS` entry to the `expected_results.txt` of the testsuite in question.

The `UNSUPPORTED` result code means that the requirements for a test are not met and so it will not be run. Currently
all tests which use the `/BE` or `/clr:pure` options are unsupported.

The `SKIP` result code indicates that a given test was explicitly skipped by adding a `SKIP` entry to the
`expected_results.txt`. A test may be skipped for a number or reasons, which include, but are not limited to: it being

> an incorrect test; it taking a very long time to run; it failing or passing for the incorrect reason.

Reviewer: Fix "number or reasons" typo and change semicolon-delimited list to normal bullet points:
```suggestion
`expected_results.txt`. A test may be skipped for a number of reasons, which include, but are not limited to:
* being an incorrect test
* taking a very long time to run
* failing or passing for the incorrect reason
```
